### PR TITLE
Fix order of language variables

### PR DESCRIPTION
**Pull request by abadger**
Previous to this change we were reading the language that we should
translate to from the `LANG` env var, disregarding other locale related
variables of higher precedence like `LC_MESSAGES` and `LC_ALL`.  This change
fixes that by getting the language to translate to from
`locale.setlocale()` which is calculated from all of the relevant user
environment variables.

* Remove the warning about using a fallback of `C.UTF-8` if the user's
  environment variables did not specify a language code valid for the
  system.

* Retrieve the language to translate to using
  `locale.setlocale("LC_MESSAGES")`.  This is better than using
  `os.environ.get("LANG")` because `setlocale()` accounts for all of the
  environment variables that are part of deciding which language the
  system gettext will use, not just `LANG`.  This aligns with various
  standards and conventions in the *NIX world.

* Remove the workarounds for setting language to `None`.  This case could
  occur when we directly tried to get the language from the environment
  variables.  With the new code we first set up the locale system by
  calling `locale.setlocale(locale.LC_ALL, locale="")` and then
  determine the langauge code using `setlocale()` a second time.
  `setlocale()` will make sure that we receive a valid language code
  after that setup has been performed.

* An example in the docstring for `configure_i18n()` should have used
 ` es_ES.UTF-8` as the locale setting.

* Reword docstring for `configure_gettext()` since we no longer use glade.

* Update unittest that checks we do not call `Locale.set()` with `None`
Also update comments and docstrings.

@ptoscano  : This is the language issue I was talking about when we met in person.  LANG ends up being used to determine the translation to use instead of LC_MESSAGES or LC_ALL, both of which should take precedence.

Also, please test this before merging.  I held off on opening the PR because I thought that `setlocale()` was returning something that the `Locale.set()` method wouldn't like but on testing it again today, I couldn't figure out what that might be... `setlocale()` returns a string in the standard `{LANGUAGE}_{COUNTRYCODE}.{ENCODING}` format which `Locale.set()` wants so I think I was wrong before but maybe I'm missing something today.

EDIT: I think what I saw before is that `selocale(locale.LC_ALL)` returns a string containing all of the locale settings, which is not what we want.  However, `setlocale(locale.LC_MESSAGES)` return only the language code used for translations, which is what we need here.

##### REPRODUCER
If you run
```
# LC_ALL=C.UTF-8 LANG=fr_FR.UTF-8 subscription-manager
```
then some messages end up translated into French instead of being the strings hardcoded into the code.  Since `LC_ALL` is supposed to take precedence over `LANG` according to POSIX, GNU coding standards, gettext documentation, etc, this is buggy behaviour.  We have been calling `subscription-manager` from within a program and screenscraping the output for a few tasks so errant translations cause us problems. (We have since worked around this by setting `LANG` and `LANGUAGE` in addition to `LC_ALL`).

### CODE REVIEW

File: src/subscription_manager/i18n.py
Comment (jirihnidek) [inline src/subscription_manager/i18n.py:63]: It is fine. We backport only security issues to RHEL-7 ATM.

File: src/subscription_manager/i18n.py
Comment (jirihnidek) [inline src/subscription_manager/i18n.py:69]: Couldn't this raise exception? The exception is not handled in this case. What is wrong with `lang = os.environ.get("LANG")`?

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Thanks for the PR. I have few comments and questions?

Is there any Bugzilla bug report related to this issue (https://bugzilla.redhat.com/)?

If not, can you please create bug report there and describe how can we reproduce this issue? It is important not only for us developers to be able to test this bug fix. It will be also important for our testing team later.

**Comment by jirihnidek** [comment]
I can see now that there are some problems with unit tests and there are also some stylish issues. Can you please fix them?

**Comment by ptoscano** [comment]
TBH, I think we should rather simplify that code, and drop all the manual handling of `os.environ` and alike in the `except` block.

Going back into the history of that file, and `configure_i18n()` in particular (and ignoring formatting/deprecation/etc changes), we have:
- commit 2131c771f417038af97b83cb626c1226018741e8: this is the one that starts the lookup into `os.environ` for the locale to use; I think the problem here was that, before this change, the `setlocale()` call in the `except` block was setting `C` (without UTF-8, thus ASCII) as locale, as done by the previous...
- ... commit 5bb36642f601518ccd916b7ff0a56f94eb7f78d6, which does something completely unrelated, and there is simply no explanation why the fallback locale was changed from `C.UTF-8` to `C` -- yay to unrelated changes :sob:
- commit a38c26806e8d5c20e47f336129a21b2113d23b47 fixes commit f1e81475431961fdbdd339efec0b6f032ed2ceae (see below), and adapts to a missing `LANG` in `os.environ`
- commit be66a6c498bef250641d24a3149274a560d26fc7 IMHO does the right thing: set `C.UTF-8` as fallback locale
- commit f1e81475431961fdbdd339efec0b6f032ed2ceae tries to use `LANG` in `os.environ`, however I think the right thing would have been to use the return value of `setlocale()`
- commit 8c4f003018a123982145f5f288d89e4b8a290e19 seems legit: deals with the missing locale (i.e. `setlocale()` raising `locale.Error`)
- commit c68905368c83e0c555d6164d4de5612cf98620a7 is the beginning of the i18n work/enablement

Considering that:
- `locale.setlocale(LC_ALL, "")` does the right thing: in the end it calls the C `setlocale()`, which Does The Right Thing when the locale string is an empty string [1]:

File: src/subscription_manager/i18n.py
Comment (abadger) [reply src/subscription_manager/i18n.py:63]: Excellent.  I will remove this comment then.

File: src/subscription_manager/i18n.py
Comment (abadger) [reply src/subscription_manager/i18n.py:69]: I believe that after we get through the try except block above, this can no longer raise an exception.

setting lang equal to the value of LANG is bad because it doesn't obey any of the other environment variables.  Of the environment variables that control translations, LANG is the fallback value.  All of the other environment variables have higher prefence and therefore if any of them are set, they should be used instead of LANG.

`setlocale(LC_MESSAGES)` used at this point, after the call to `setlocale(LC_ALL, "")`, should return the proper value for translations taking into account all of the relevant environment variables.  The other option would be to code our own set of fallbacks that follow the standards, examining LANGUAGE, LC_ALL, LC_MESSAGES, and LANG for the relevant language code to use for translations.

**Comment by abadger** [comment]
> ```python
>     new_locale: str
>     try:
>         new_locale = locale.setlocale(category=locale.LC_ALL, locale="")
>     except locale.Error:
>         os.environ["LC_ALL"] = "C.UTF-8"
>         new_locale = locale.setlocale(locale.LC_ALL, locale="C.UTF-8")
> 
>     configure_gettext()
>     Locale.set(new_locale)
> ```
This looks good to me.  I will update the PR.

**Comment by abadger** [comment]
PR Updated.  The one change I made to ptoscano's proposed code was to get the language code from `setlocale(locale.LC_MESSAGES)` instead of `setlocale(locale.LC_ALL)`.  This is because LC_ALL returns a string that contains all of the locale settings.  This is not what we need.  (Also, the LC_MESSAGES category is precisely the setting that we need to look at to determine the language to translate into.  One of the bugs in the former code was buggy because it tried to query the more all-encompassing LANG variable instead of the specific LC_MESSAGES which could override that so querying LC_MESSAGES will avoid making that same mistake.

My commits should also have fixed the unittests now.  There's also the option to remove that particular unittest, `test_configure_i18n_lang_none()` as None isn't really a possibility from the output of `setlocale()` but I decided to leave it in to prevent regressions.  Let me know if you would rather have it removed.

**Comment by abadger** [comment]
Oh, about the reproducer... If you run
```
$ LC_ALL=C.UTF-8 LANG=fr_FR.UTF-8 subscription-manager
```
then some messages end up translated into French instead of being the strings hardcoded into the code.  Since `LC_ALL` is supposed to take precedence over `LANG` according to POSIX, GNU coding standards, gettext documentation, etc, this is buggy behaviour.  We have been calling `subscription-manager` from within a program and screenscraping the output for a few tasks so errant translations cause us problems. (We have since worked around this by setting LANG and LANGUAGE in addition to LC_ALL).

**Comment by ptoscano** [comment]
Sounds good, it will need a bit more of testing.

Can you please squash the commits, and create a single one with explanation and test fixes? I don't think having the intermediate changes is useful, also they break bisection.

**Comment by abadger** [comment]
> Can you please squash the commits, and create a single one with explanation and test fixes? I don't think having the intermediate changes is useful, also they break bisection.

Done.  Squashed commits are definitely better before a change is merged.

**Comment by ptoscano** [comment]
Sorry for the delay on this; it generally LGTM, so can you please rebase it, so we get updated CI, and I can trigger QE tests?

**Comment by ptoscano** [comment]
@abadger friendly ping? ;)

**Comment by abadger** [comment]
@ptoscano Thanks for the reminder!  Rebased and pushed.  (Note: My availability will be sporadic almost all of June.  Things to take care of prior to travelling, PTO, and the conversions as a service meetings are going to mean I'm not working regular hours [or as much] throughout that month.)

**Review by ptoscano — APPROVED** [review]
Thanks for the rebase! The Fedora Rawhide failure is not related, so it is not a blocker.

LGTM now, thanks! :+1: There were no regressions in our QE test suite. I'll leave it to @jirihnidek for a second review, since he had notes in the past.

**Review by jirihnidek — APPROVED** [review]
LGTM

**Comment by abadger** [comment]
Thanks all!

**Pull request merged**
